src/compiler/emitter.ts

```diff
@@ -5657,6 +5869,91 @@ var __param = (this && this.__param) || function (paramIndex, decorator) {
                 }
             }
 
+            function emitJsxElement(node: JsxElement|JsxSelfClosingElement) {
+                switch (compilerOptions.jsx) {
+                    case JsxEmit.React:
+                        jsxEmitReact(node);
+                        break;
+                    // Fall back to preserve if None was specified (we'll error earlier)
+                    default:
+                        jsxEmitPreserve(node);
+                        break;
+                }
+            }
+
+            function trimReactWhitespace(node: JsxText): string {
+                // Could be empty string, do not use !node.formattedReactText
+                if (node.formattedReactText !== undefined) {
+                    return node.formattedReactText;
+                }
+
+                let lines: string[] = [];
+                let text = getTextOfNode(node);
+                let firstNonWhitespace = 0;
+                let lastNonWhitespace = -1;
+
+                for (var i = 0; i < text.length; i++) {
+                    var c = text.charCodeAt(i);
+                    if (c === CharacterCodes.lineFeed || c === CharacterCodes.carriageReturn) {
+                        if (firstNonWhitespace !== -1 && (lastNonWhitespace - firstNonWhitespace + 1 > 0)) {
+                            lines.push(text.substr(firstNonWhitespace, lastNonWhitespace - firstNonWhitespace + 1));
+                        }
+                        firstNonWhitespace = -1;
+                    }
+                    else if (!isWhiteSpace(c)) {

```

Comment:
This function's naming is inconsistent with other identifiers using the term "whitespace"